packages/tpetra/core/example/BlockCrs/MeshDatabase.hpp

```diff
@@ -0,0 +1,335 @@
+//
+//          Tpetra: Templated Linear Algebra Services Package
+//                 Copyright (2008) Sandia Corporation
+//
+// Under the terms of Contract DE-AC04-94AL85000 with Sandia Corporation,
+// the U.S. Government retains certain rights in this software.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY SANDIA CORPORATION "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Michael A. Heroux (maherou@sandia.gov)
+//
+// ************************************************************************
+// @HEADER
+
+#ifndef __MESHDATABASE_HPP__
+#define __MESHDATABASE_HPP__
+
+#include <iostream>
+#include <sstream>
+
+#include "typedefs.hpp"
+#include "Teuchos_Comm.hpp"
+
+namespace BlockCrsTest {
+
+  template<typename T> 
+  KOKKOS_INLINE_FUNCTION
+  static T
+  get_block_crs_entry(const LO i, const LO j, const LO k, 
+                      const LO diff_i, const LO diff_j, const LO diff_k, 
+                      const LO l0, const LO l1) {
+    // any reasonable deterministic number to fill diagonal blocks 
+    // inside of the block, put some weight on diagonals
+    // make off diagonal blocks a bit smaller than the diagonals
+    // add epsilon to remove clean zeros
+    return ( ( (i%13)/14.0 + (j%17)/12.0 + (k%19)/10.0 ) +  // any number to fill diagonal blocks
+             ( l0 == l1 ? 10.0*(l0%10)/10.0 : ((l0+l1)%10)/10.0) + 
+             ( diff_i*((i-j)%4)/18.0 + diff_j*((i-j)%7)/24.0 + diff_k*((i-j)%5)/16.0) + 
+             ( 1.0e-7 ) );
+  }
+
+  template<typename T> 
```

Comment:
It is because our current interest is on double precision only. But later we can add complex test too. I just prepare for the code for possible extension.